### PR TITLE
[Feat] Nonautoregressive Methods

### PR DESCRIPTION
**Pull request by cbhua**
## Description

Contributed by @Furffico, @tycbony, and @henry-yeh from DeepACO and GLOP. Adding new support for non-autoregressive (NAR) methods, including edge embedding methods for various environments, anisotropic GNN encoder with edge-gating mechanism, non-autoregressive decoder, and related utils functions. 

## Motivation and Context

NAR methods are an important extension for RL4CO model support, aside from various previous autoregressive methods. 

## Types of changes

- [x] New feature (mostly just adding new functionality)
- [x] Example (update in the folder of examples)
- [x] Breaking change (there are some changes to basic classes, even though most of them are simple code-level optimization, just be careful about hidden bugs)

## Checklist

- [ ] This change requires an extension of the documentation.
- [x] This change has provided a testing.

### CODE REVIEW

**Review by cbhua — COMMENTED** [review]
Great code, nice work.

File: rl4co/models/nn/env_embeddings/edge.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import torch
+import torch.nn as nn
+
+from torch import Tensor
+
+try:
+    from torch_geometric.data import Batch, Data
```
[the remark this answers is not in the document]
Comment (fedebotu) [reply rl4co/models/nn/env_embeddings/edge.py:7]: I think there are two ways. Honestly, I think the current version with the import error is on the right direction! But we can make some improvements.

The first way would be to enforce `torch_geometric` installation. But this is not optimal for some users because of pretty strict requirements.
The second way (and IMO preferable) is not to enforce installation (as done now) but raise error _only_ when we try to initialize the module. We could instead import `Batch` and `Data` as `None`. Then, if the module is initialized, raise the `ImportError`.
We can follow this example: [1) import with None ](https://github.com/Dao-AILab/flash-attention/blob/6c9e60de566800538fedad2ad5e6b7b55ca7f0c5/flash_attn/modules/mlp.py#L16)and [2) raise error upon initialization](https://github.com/Dao-AILab/flash-attention/blob/6c9e60de566800538fedad2ad5e6b7b55ca7f0c5/flash_attn/modules/mlp.py#L70) )


---

PS: let's also make a `all` requirements for installing everything as [here](https://stackoverflow.com/questions/75490231/pip-pyproject-toml-can-optional-dependency-groups-require-other-optional-depend).

**Review by fedebotu — APPROVED** [review]

File: rl4co/models/nn/env_embeddings/edge.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Furffico) [reply rl4co/models/nn/env_embeddings/edge.py:7]: > The second way (and IMO preferable) is not to enforce installation (as done now) but raise error _only_ when we try to initialize the module. We could instead import `Batch` and `Data` as `None`. Then, if the module is initialized, raise the `ImportError`. 

The second way seems more appropriate. I'll try with that one.

> PS: let's also make a `all` requirements for installing everything as [here](https://stackoverflow.com/questions/75490231/pip-pyproject-toml-can-optional-dependency-groups-require-other-optional-depend).

Does `all` need to include development dependencies like `testing` and `dev`?

Moreover, regarding optional dependencies, should we rename `graph` to `nar` to indicate that it is a dependency for the optional feature `nar`?

[the remark this answers is not in the document]
Comment (cbhua) [reply rl4co/models/nn/env_embeddings/edge.py:7]: I think `all` doesn't have to include `dev` and `testing` since these two parts of dependencies are only used for developers, and only called from RL4CO repository source code. 
```
testing = ["pytest", "pytest-cov"]
dev = ["black", "ruff", "pre-commit>=3.3.3"]
```
As for the `graph`, we also have the [GCN encoder](https://github.com/ai4co/rl4co/blob/main/rl4co/models/nn/graph/gcn.py) requires `torch_geometric`. Maybe we could keep it name as `graph` in the current state and later if we have good ideas to make it clearer we could update then.

File: rl4co/models/nn/env_embeddings/edge.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    def _cost_matrix_to_graph(self, batch_cost_matrix: Tensor, init_embeddings: Tensor):
+        """Convert batched cost_matrix to batched PyG graph, and calculate edge embeddings.
+
+        Args:
+            batch_cost_matrix: Tensor of shape [batch_size, n, n]
+            init_embedding: init embeddings
+        """
+        graph_data = []
+        for index, cost_matrix in enumerate(batch_cost_matrix):
```
[the remark this answers is not in the document]
Comment (Furffico) [reply rl4co/models/nn/env_embeddings/edge.py:85]: https://github.com/ai4co/rl4co/blob/567d58a21c3210555af4741594889efb1c61854a/rl4co/models/nn/env_embeddings/edge.py#L103-L104

[`torch_geometric.data.Batch`](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.data.Batch.html) achieves the conversion from a batch of graphs to a huge disconnected graph, but `from_data_list` is the only way to instantiate a `Batch` object.

If your idea is to create the huge graph directly without the conversion, we need to get rid of `torch_geometric.data.Batch` and rewrite [`sparsify_graph`](https://github.com/ai4co/rl4co/blob/567d58a21c3210555af4741594889efb1c61854a/rl4co/utils/ops.py#L158-L194) into a batched version, which might be possible with [masked tensors](https://pytorch.org/docs/stable/masked.html) replacing [`fill_diagonal_`](https://pytorch.org/docs/stable/generated/torch.Tensor.fill_diagonal_.html).

**Comment by Furffico** [comment]
I have modified the code following [the review above](#pullrequestreview-1911485186). Please review the changes.

**Review by cbhua — COMMENTED** [review]
Great work, thanks!
